### PR TITLE
Fix StyleCop rule SA1129 ("Do not use default value type constructor")

### PR DESCRIPTION
**Pull request by stephentoub**
Corresponds to coreclr change https://github.com/dotnet/coreclr/pull/27716

### CODE REVIEW

File: src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/FailedProxyCache.cs [hunk 1]
```diff
@@ -36,7 +36,7 @@ internal sealed class FailedProxyCache
 
         // This lock can be folded into _nextFlushTicks for space optimization, but
         // this class should only have a single instance so would rather have clarity.
-        private SpinLock _flushLock = new SpinLock();
+        private SpinLock _flushLock = new SpinLock(enableThreadOwnerTracking: false); // mutable struct; do not make this readonly
```
Comment (sharwell) [inline src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/FailedProxyCache.cs]: I really wish non-copyable value types was a thing (or at minimum, the compiler provided information about all copies so an analyzer could make it effectively non-copyable).

**Comment by bartonjs** [comment]
It looks syntactically valid... but why?

**Comment by stephentoub** [comment]
> why?

A few reasons. First and foremost, because of concerns raised (and that I buy and have had similar thoughts about in the past) that using the struct parameterless ctors can be misleading, e.g. thinking that `new Guid()` is the same as `Guid.NewGuid()` when in fact they are very different. Second, in my experience it makes code review harder, as you need to spend time looking up whether types bring constructed are allocating or not. Third, it's simpler code. Fourth, IL allows structs to have parameterless ctors and C# is considering adding support for that in the future; at that point, I think there's a lot of value in being able to see when you're calling one of those vs just default initializing.

**Comment by bartonjs** [comment]
Does that mean that we're banning object initialization for the mutable value types?

```C#
RSAParameters rsaParams = new RSAParameters
{
    Modulus = someVal,
    Exponent = probablyF4,
};
```

They still say `new` in them, after all.  (I'd like the answer to be "no", but I'm wondering where the consistency is)

**Comment by stephentoub** [comment]
> Does that mean that we're banning object initialization for the mutable value types?

No. It's a balance.

**Comment by sharwell** [comment]
Those cases are not reported for SA1129.

https://github.com/DotNetAnalyzers/StyleCopAnalyzers/blob/59adc837b84e4d8885010ffec2c5c402db66e745/StyleCop.Analyzers/StyleCop.Analyzers.Test/ReadabilityRules/SA1129UnitTests.cs#L64-L65

**Comment by bartonjs** [comment]
> It's a balance.

Fair enough.

I'm personally not a fan of letting C# declare default ctors on value types, because it's a concept-breaking change.  But I can support heading the ambiguity off at the pass.

**Review by bartonjs — APPROVED** [review]
s/Approve/Acquiesce/ :smile:

**Comment by stephentoub** [comment]
> s/Approve/Acquiesce/

I saw you use some C#8 syntax in your recent large PR. Welcome :)

**Review by joperezr — APPROVED** [review]
LGTM. Are we waiting on something else or can this one get merged?

**Comment by stephentoub** [comment]
I'm waiting for the mirror PR to land first.

**Review by safern — APPROVED** [review]

**Pull request merged**